### PR TITLE
Feature/CSVB-19421 - Include activity event processing

### PR DESCRIPTION
**Pull request by waringr**
Includes Activity event processing.

### CODE REVIEW

File: src/services/activities-document-conversion.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        throw err;
+    } finally {
+        activityConnection.release();
+    }
+
+    return upsertResults;
+};
+
+const deleteActivity = async (activity: Activity): Promise<void> => {
+    throw new Error("deleting tech record documents is not implemented yet");
```
Comment (ChrisPeska-BJSS) [inline src/services/activities-document-conversion.ts:68]: correct the error message so it matches Activity

File: src/services/table-details.ts [hunk 2]
```diff
@@ -385,6 +385,29 @@ export const TEST_RESULT_TABLE: TableDetails = {
...
+
+export const WAIT_REASON_TABLE: TableDetails = {
+    tableName: "wait_reason",
+    columnNames: [
+        "activity_id",
+        "note"
+    ]
+};
+
 export const allTables = (): TableDetails[] => {
```
Comment (ChrisPeska-BJSS) [inline src/services/table-details.ts:411]: I am not sure if is used anywhere, should it contain WAIT_REASON_TABLE as well ?

File: src/services/activities-document-conversion.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    const upsertResults: ActivitiesUpsertResult[] = [];
+    const activityConnection = await pool.getConnection();
+
+    try {
+        await activityConnection.execute("SET SESSION TRANSACTION ISOLATION LEVEL READ COMMITTED");
+
+        const testStationId = await upsertTestStation(activityConnection, activity);
+        const testerId = await upsertTester(activityConnection, activity);
+
+        debugLog(`upsertTechRecords: Upserting tech record..`);
```
Comment (ChrisPeska-BJSS) [inline src/services/activities-document-conversion.ts:31]: Correct debug output message for Activity

File: src/models/activity.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    waitReason?: string[];
+}
+
+export const parseActivity = (image: DynamoDbImage): Activity => {
+    debugLog("Parsing activity...");
+
+    return {
+        id: image.getString("id"),
+        activityType: image.getString("activityType"),
+        activityDay: image.getString("activityDay"),
```
Comment (ChrisPeska-BJSS) [inline src/models/activity.ts:28]: this attribute is represented by DATE type in the database

Comment (waringr) [reply src/models/activity.ts:28]: That's just YYYY-MM-DD isn't it?

Comment (waringr) [reply src/models/activity.ts:28]: I think that's what it comes in as, so it can stay as string, cause there's no conversion to do

File: src/services/activities-document-conversion.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        connection
+    );
+
+    debugLog(`upsertTestResults: Upserted tester (ID: ${response.rows.insertId})`);
+
+    return response.rows.insertId;
+};
+
+const upsertWaitReasons = async (connection: Connection, activityId: string, activity: Activity): Promise<number[]> => {
+    debugLog(`upsertTechRecords: Upserting axle spacings (activity-id: ${activityId})...`);
```
Comment (ChrisPeska-BJSS) [inline src/services/activities-document-conversion.ts:108]: correct debug log message

File: src/services/activities-document-conversion.ts [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    if (!activity.waitReason) {
+        debugLog(`upsertTechRecords: no wait reasons present`);
```
Comment (ChrisPeska-BJSS) [inline src/services/activities-document-conversion.ts:111]: correct debug log message

File: src/services/table-details.ts (hunk 2 above)
Comment (waringr) [reply src/services/table-details.ts:411]: Added wait reason table to this list

File: src/models/activity.ts (hunk 4 above)
Comment (ChrisPeska-BJSS) [reply src/models/activity.ts:28]: From API Schema
`activityDay:
  type: string
  format: date
  nullable: true
`
All dates are of the string type in test result / tech records / activity, if you leave it as they are there will be implicit type casting as in database this field is DATE.

File: src/services/table-details.ts [hunk 7]
```diff
@@ -385,6 +385,29 @@ export const TEST_RESULT_TABLE: TableDetails = {
...
+        "endTime",
+        "notes"
+    ]
+};
+
+export const WAIT_REASON_TABLE: TableDetails = {
+    tableName: "wait_reason",
+    columnNames: [
+        "activity_id",
+        "note"
```
Comment (ChrisPeska-BJSS) [inline src/services/table-details.ts:407]: This field is called `reason`.

File: .gitmodules [hunk 8]
```diff
@@ -1,4 +1,4 @@
 [submodule "cvs-nop"]
 	path = cvs-nop
 	url = https://github.com/dvsa/cvs-nop
-	branch = CVSB-19155
+	branch = feature/CVSB-19421
```
Comment (JonLazariniDD) [inline .gitmodules:4]: I think we might need to come up with a way to inject our branch as an environmental variable so it picks up the changes of the relevant branch, or changes we want to make.
I am thinking of cases where the developer or remote has the cvs-nop updated (and their schema) which could lead to situation where the DB loaded locally is not what we expect.
What about having a script that generates this file and injects the branch just after we have cloned and installed the repo?
Then once the file is generated we would just have to run git submodule update --init --recursive ?

File: src/models/activity.ts (hunk 4 above)
Comment (JonLazariniDD) [reply src/models/activity.ts:28]: Dates are strings formatted as ISO8601 in our micro services with cases where have legacy data (that we are aware of) which could contain up to 6 digits for the ms. (calculation of expiry date for certificates)
```js
 public static isValidDate(
    input: string | Date | number | undefined
  ): boolean {
    return (
      input !== undefined &&
      (moment(input, "YYYY-MM-DDTHH:mm:ss.SSSSSSZ", true).isValid() || // Legacy format
        moment(input, "YYYY-MM-DDTHH:mm:ss.SSSZ", true).isValid() || //  Test result API format
        moment(input, "YYYY-MM-DD", true).isValid()) && // Tech Record dates e.g. regnDate
      moment(input).isAfter(new Date(0))
    );
  }
  ```

File: src/models/activity.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
+import {DynamoDbImage, parseStringArray} from "../services/dynamodb-images";
+import {debugLog} from "../services/logger";
+
+export interface Activity {
+    id?: string;
+    activityType?: string;
+    activityDay?: string;
```
Comment (JonLazariniDD) [inline src/models/activity.ts:7]: What is this `activityDay` please? I am not sure we are capturing this as part of the activity.
I couldn't find it in the 'dynamo' schema
```js
export const ActivitySchema = Joi.object().keys({
    parentId: Joi.string().optional(),
    activityType: Joi.any().only([ activitiesTypes ]).required(),
    testStationName: Joi.string().required(),
    testStationPNumber: Joi.string().required(),
    testStationEmail: Joi.string().email().required(),
    testStationType: Joi.any().only([ stationTypes ]).required(),
    testerName: Joi.string().min(1).max(60).required(),
    testerStaffId: Joi.string().required(),
    testerEmail: Joi.any().when("activityType", {
        is: ActivityType.VISIT,
        then:  Joi.string().email().required(),
        otherwise: Joi.any().forbidden(),
    }),
    startTime: Joi.string().optional(),
    endTime: Joi.string().optional().allow(null),
    waitReason: Joi.array().items([ waitReasons ]).optional(),
    notes: Joi.string().allow(null)
});
```

Also `testStationType` is not in there so i assume we are not interested in it?.
I can't see the `parentId` either which are required for all activities type except visit (wait and unaccountable time)

**Review by JonLazariniDD — COMMENTED** [review]
I think the infrastructure is down but I wanted to look at the new tables we have created for those and couldn't find them so I might need your help with this when you have some time please if it is fine with you.

File: src/models/activity.ts (hunk 9 above)
Comment (ChrisPeska-BJSS) [reply src/models/activity.ts:7]: `activityDay` is present in DynamoDB tables cvs-[integration|develop]-activities and samples have it populated, it is present in API schema. `testStationType` is present in samples too and in your code excerpt - we need it for fingerprint matching in database. Relationship/logic between attributes are not relevant for cvs-tsk-update-store to my mind (as long as it does not affect the way we store the data) - we store Activity essentially as-is.

File: src/models/activity.ts (hunk 4 above)
Comment (ChrisPeska-BJSS) [reply src/models/activity.ts:28]: This is good to know, as it was a black-box before for us. API schema is sketchy sometimes and does not explicitly describe what format is expected. I would insist on explicit casting instead of treating it as a string.

File: src/models/activity.ts (hunk 9 above)
Comment (JonLazariniDD) [reply src/models/activity.ts:7]: Thanks! I will create ticket for this so we can update our schemas and also move the api specs in the repo assuming we don't have secrets, etc.. in them.
There is probably a use case we will need to discuss with the team  as per my understanding:
An activity/visit can be of different types: visit, unaccountable time, wait time and I couldn't whether we are capturing the following relationship:
visit activity can have 'sub-activities' which can be of type wait or unaccountable time where a parentId is needed and the parentId is activityId of the visit activity.
basically when starting a visit in the mobile app this will create a visit and unaccountable time (1), wait times (2) are added if the user does nothing after a certain amount of time (2), or just close a test (1) for whatever reasons.

Comment (waringr) [reply src/models/activity.ts:7]: @ChrisPeska-BJSS Can you answer this one?

Comment (ChrisPeska-BJSS) [reply src/models/activity.ts:7]: Thanks! That is piece of good information. I assumed there must be hierarchic relationship within data - the presence of `parentId`. We can disregard this and store data in `Activity` table only. It all depends how this data is going to be utilised. At the moment the database is hosted on AWS Aurora (MySql), which unlike other RDBMSes poorly supports hierarchic recursive queries (recursive common table expression or dedicated sytnax e.g. Oracle Connect By). It will be surmountable to query and extract hierarchic information, but it is not going to be straight-forward. I need to have @SanjeetBJSS steer on it ? Possibly, we are going to keep it as it is and store everything in one table.

Having said that should the model have 'enum' types ?

Comment (bristol-mike-jones) [reply src/models/activity.ts:7]: The Activities table stores two types of data.  You get the Visit and then you get Wait time & Unaccountable time.  These last two are very similar and relate to a parent visit.  You could model these in separate tables if you want to avoid sparsely populated tables.  Otherwise, you could merge them together.

The parent relationship is only a single level.  i.e. a parent cannot have a parent.  So while you might want to join the table to itself to get visits and the wait/unaccounted information for it - you wouldn't want to tree walk.

Comment (waringr) [reply src/models/activity.ts:7]: Need to add indexes on activityId and parentId so we can link the activities to themselves via those columns @ChrisPeska-BJSS

File: src/models/activity.ts (hunk 4 above)
Comment (waringr) [reply src/models/activity.ts:28]: Remove `activity_day` column from DB and remove references from `update-store` as well.

File: .gitmodules (hunk 8 above)
Comment (JonLazariniDD) [reply .gitmodules]: As discussed, created a ticket https://jira.dvsacloud.uk/browse/CVSB-19447 and resolving

File: README.md [hunk 10]
```diff
@@ -12,21 +13,24 @@ See full conversion procedure in sections below.
 This Lambda currently supports the following conversions:
 
 | Source (DynamoDB)   | Destination (Aurora) |
-|---------------------|----------------------|
+| ------------------- | -------------------- |
```
Comment (bristol-mike-jones) [inline README.md:16]: If we are going to say which DynamoDB tables this processes - it needs to include Activities!

File: README.md [hunk 11]
```diff
@@ -104,41 +118,43 @@ Rough ordering - may not follow code exactly.
...
+  - `vehicle_class`
+  - `preparer`
+  - `identity`
+  - `fuel_emission`
+  - `test_type`
+  - `test_result`
+  - `defect`
+  - `location`
+  - `test_defect`
+  - `custom_defect`
```
Comment (bristol-mike-jones) [inline README.md:157]: Again, if we are going to document this here - it needs to include activities

File: tests/unit/models/activity.unitTest.ts [hunk 12]
```diff
@@ -0,0 +1,28 @@
+import { parseActivity, Activity } from "../../../src/models/activity";
+import {DynamoDbImage} from "../../../src/services/dynamodb-images";
+import {default as activityJson} from "../../resources/dynamodb-image-activity.json";
+import {castToImageShape} from "../../utils";
+
+describe("parseTechRecordDocument()", () => {
+    it("should successfully parse a DynamoDB image into a TechRecordDocument", () => {
```
Comment (bristol-mike-jones) [inline tests/unit/models/activity.unitTest.ts:7]: Typo's from copying the TechRecord code ;)
